warehouse/integrations/deltalake-native/deltalake_native.go

```diff
@@ -0,0 +1,1367 @@
+	d.Warehouse = warehouse
+	d.Namespace = warehouse.Namespace
+	d.Uploader = uploader
+	d.ObjectStorage = warehouseutils.ObjectStorageType(warehouseutils.DELTALAKE, warehouse.Destination.Config, d.Uploader.UseRudderStorage())
+
+	db, err := Connect(d.credentials())
+	if err != nil {
+		return fmt.Errorf("connecting: %w", err)
+	}
+
+	d.DB = db
+
+	return nil
+}
+
+// TestConnection tests the connection to the warehouse
+func (d *Deltalake) TestConnection(warehouse model.Warehouse) error {
+	d.Warehouse = warehouse
+
+	db, err := Connect(d.credentials())
+	if err != nil {
+		return fmt.Errorf("connecting: %w", err)
+	}
+
+	d.DB = db
+	defer func() { _ = d.DB.Close() }()
+
+	ctx, cancel := context.WithTimeout(context.TODO(), d.ConnectTimeout)
+	defer cancel()
+
+	if err = d.DB.PingContext(ctx); errors.Is(err, context.DeadlineExceeded) {
+		return fmt.Errorf("connection timeout: %w", err)
+	}
+	if err != nil {
+		return fmt.Errorf("pinging: %w", err)
+	}
+
+	return nil
+}
+
+// Cleanup cleans up the warehouse
+func (d *Deltalake) Cleanup() {
+	if d.DB != nil {
+		d.dropDanglingStagingTables()
+		_ = d.DB.Close()
+	}
+}
+
+// CrashRecover crash recover scenarios
+func (d *Deltalake) CrashRecover(warehouse model.Warehouse) error {
+	d.Warehouse = warehouse
+	d.Namespace = warehouse.Namespace
+
+	db, err := Connect(d.credentials())
+	if err != nil {
+		return fmt.Errorf("connecting: %w", err)
+	}
+
+	d.DB = db
+	defer func() { _ = d.DB.Close() }()
+
+	d.dropDanglingStagingTables()
+
+	return nil
+}
+
+// IsEmpty checks if the warehouse is empty or not
+func (*Deltalake) IsEmpty(model.Warehouse) (bool, error) {
+	return false, nil
+}
+
+// LoadUserTables loads user tables
+func (d *Deltalake) LoadUserTables() map[string]error {
+	return d.loadUserTables()
+}
+
+// LoadTable loads table for table name
+func (d *Deltalake) LoadTable(tableName string) error {
+	_, err := d.loadTable(tableName, d.Uploader.GetTableSchemaInUpload(tableName), d.Uploader.GetTableSchemaInWarehouse(tableName), false)
+	if err != nil {
+		return fmt.Errorf("loading table: %w", err)
+	}
+
+	return nil
+}
+
+// LoadIdentityMergeRulesTable loads identifies merge rules tables
+func (*Deltalake) LoadIdentityMergeRulesTable() error {
+	return nil
+}
+
+// LoadIdentityMappingsTable loads identifies mappings table
+func (*Deltalake) LoadIdentityMappingsTable() error {
+	return nil
+}
+
+// DownloadIdentityRules download identity rules
+func (*Deltalake) DownloadIdentityRules(*misc.GZipWriter) error {
+	return nil
+}
+
+// GetTotalCountInTable returns the total count in the table
+func (d *Deltalake) GetTotalCountInTable(ctx context.Context, tableName string) (int64, error) {
+	var (
+		total int64
+		err   error
+		query string
+	)
+
+	query = fmt.Sprintf(`
+		SELECT COUNT(*) FROM %[1]s.%[2]s;
+	`,
+		d.Namespace,
+		tableName,
+	)
+
+	err = d.DB.QueryRowContext(ctx, query).Scan(&total)
+
+	if err != nil {
+		if strings.Contains(err.Error(), schemaNotFound) {
```

Comment:
I believe we should try adding tests in all such situations to make sure that the error returned is indeed what we expected.

Reply:
Right now we are waiting for error types to get added to the databricks SQL go driver.
https://github.com/databricks/databricks-sql-go/pull/117